### PR TITLE
SPort Passthrough telemetry via MAVLink (WIP, do not merge)

### PR DESCRIPTION
**Pull request by olliw42**
@yaapu FYI

this here is a 1st attempt at something which I carry with me since a while, namely to stream native Frsky SPort passthrough telemetry packets via mavlink. The corresponding mavlink message shall host an array of sport packets, somewhat similar to the multi passthrough packet.

The motivation is at least 4 fold:
- On bandwidth limited mavlink radio links, sending all the stuff which one wants to have for the typically telemetry  consumes quite some bandwidth in the rx->tx direction. While it obviously can be adapted and shrunk down, a typical order of magnitude would be ca 1500 bytes/sec. However, the passthrough telemetry protocol also essentially provides all the basic data, but in a much more compressed way (see also below). Therefore, if that data could be send instead of the original mavlink messages much bandwidth could be saved - or, also often much desired, more fluent telemetry updates could be achieved.
- The passthrough protocol has been over the years highly optimized to provide what the majority of users want to get, and it in fact provides info which is not so easy to get via mavlink or may not even be available. Examples would be some parameter values such as BATT capacity values and other. They of course can be retrieved with mavlink too, however with significant burden both implementation-wise and link-wise. An example of info not consistently available with mavlink is the is_flying flag. Therefore, passthrough via mavlink would significantly easy the situation.
- A passthrough via mavlink message could serve as a viable, and in many situations maybe better, alternative  to the HIGH_LATENCY messages. In fact, these messages are obviously very specific to very specific use cases, and hardly can serve as more general telemetry messages for high-latency situations. This is so because much info is either not provided while other info is duplicated in the two versions of HIGH_LATENCY. A passthrough mavlink message however is very generic and hence can carry varying and different types of data. Moreover, it is very future proof as the link does not have to be modified upon additions to the passthrough protocol.
- This argument is due to @yaapu. Currently, mavlink-to-passthrough converters (such as Mav2PT) need to be adapted or modified whenever a change or addition to the passthrough protocol is made. This obviously makes maintenance and upgrading a significant task, as there are so many parties involved. With such a message only e.g. ArduPilot and the yappu app need to be adapted, but not all the things which are in between. 

The 1st implementation shared here has been demonstrated to work in combination with mLRS (https://github.com/olliw42/mLRS/tree/dev-crsfpassthruarray). The following pictures shall demonstrate this. The first two pictures show the mavlink data which is streamed on serial2 (which is connected to mLRS), which should demonstrated that there is really  not much data send. The last two pictures show that much data is however received by the yappu telemetry app (in fact, some of the data would not be available with just a mavlink-to-passthrough conversion).

MissionPlanner reports 327 bytes/sec for the TUNNEL messages send at 4 Hz. This is obviously much less than the typical ca 1500 bytes/sec.

crsfpassthruarray-sr2params

crsfpassthruarray-snif

IMG_20220922_184445

IMG_20220922_184515

The current implementation has however a number of pain points, which would have to be overcome:
- It uses the TUNNEL mavlink message. While this is technically fine, I think one should better define a new specific mavlink message. An attempt can be found in bp_mavlink/development.xml, called FRSKY_PASSTHROUGH_ARRAY. The main reason is that it would not possible with the current mavlink standard to request this message to be streamed. A dedicated message would overcome this limitation. I used the TUNNEL message since it is there on all involved parties. While it should IMHO be done, bringing up a new mavlink message is a tedious process. For development TUNNEL shall be ok. 
- The streaming is currently linked into Extra2, together with VFR_HUD. It would probably be more appropriate, and more desirable to have an extra streaming channel for that. At least I would like this much better. I choose Extra2 solely for the reason that it had the least number of messages currently.
- I had to duplicate much of the passthrough code. The existing classes could not be used with significant quirks - at least as far as I understand the code. For instance, AP_Frsky_SPort_Passthrough could not be new-ed, since it would not work with a nullptr for _port. And so on. However, separating out the protocol parts could actually lead to a clean up and better structure of the existing ArduPilot code, as the passthrough protocol stuff is currently spread over several files & classes.
- there are //OW's around, which really help me to navigate through the code, but which of course have to go. Pl be gentle with me here.
- There are probably many more pain points, which I miss, and which we hopefully can resolve.

The most important question to be settled for now is IMHO if this concept has any future in ArduPilot at all. (if not I am happy to go with my BetaPilot, but let's see LOL).

Olli

### CODE REVIEW

**Comment by olliw42** [comment]
argh ... something must have went wrong ... it is not on top of copter4.3, screwing up what should be a relatively compact and easy to follow pr ... no idea how to correct this. sorry folks.

**Comment by yaapu** [comment]
Hi @olliw42 nice attempt :-)
from my "personal and selfish" point of view this would simplify a lot supporting setups like MavToPT and the like.
One open point is a good way to reuse the passthrough code, right now we do that in the FPort/FPort2 and CRSF telemetry implementations, basically you call **AP_Frsky_Telem::get_telem_data(&telem_data.packet, packet_count, 1)** 
The frsky telemetry is a singleton so you cannot have frsky telemetry on serial port AND an external consumer but in your case this should not be a problem, if you want to tunnel frsky telemetry thru mavlinbk you do not need it on the RC link

https://github.com/ArduPilot/ardupilot/blob/f9bc7e44f85266abd66fe59a93d0e75de3a7b645/libraries/AP_Frsky_Telem/AP_Frsky_Telem.cpp#L120-L127
Whenever you call that method you actually invoke the frsky WFQ packet scheduler so you get the packets meant to be sent based on the global priorities defined here 
https://github.com/ArduPilot/ardupilot/blob/f9bc7e44f85266abd66fe59a93d0e75de3a7b645/libraries/AP_Frsky_Telem/AP_Frsky_SPort_Passthrough.cpp#L121-L150

Check here to have an idea on how the multi frame packet is built based on the requested size
https://github.com/ArduPilot/ardupilot/blob/f9bc7e44f85266abd66fe59a93d0e75de3a7b645/libraries/AP_Frsky_Telem/AP_Frsky_SPort_Passthrough.cpp#L815-L857

**Comment by yaapu** [comment]
To prevent corrupting RC telemetry you would need to check if custom CRSF telemetry or FrSKY passthru telemetry is enabled

**Comment by yaapu** [comment]
Since you'll have exclusive use of the frsky telemetry singleton you can tune the individual streams and perhaps disable the unwanted ones, I do this in the CRSF telemetry implementation for I do not need status text messages sent as frsky frames, an example of individual tuning of the streams is here 
https://github.com/ArduPilot/ardupilot/blob/f9bc7e44f85266abd66fe59a93d0e75de3a7b645/libraries/AP_RCTelemetry/AP_CRSF_Telem.cpp#L123-L127

**Comment by olliw42** [comment]
> To prevent corrupting RC telemetry you would need to check if custom CRSF telemetry or FrSKY passthru telemetry is enabled, so basically FPort, FPort2, SPort and CRSF custom telemetry

I think that's no problem.
First off, I think neither should be generically enabled if this is used. If any is, I think this should not impact them, as the routines I have copied over are "state-less", and if they were not I have made them state-less. That's why for instance uint32_t calc_velandyaw(bool airspeed_enabled, bool send_airspeed); has these bools now. None of them (should) use any Frsky/Crsf class internals. So, IMHO no cross effect. At least as much as I understand it LOL.

The idea would also be that the original code, like that in AP_Frsky_SPort_Passthrough.cpp could/should use the introduced AP_Frsky_SPort_Protocol class. For instance AP_Frsky_SPort_Passthrough::calc_velandyaw(void) could become (where _protocol would be created in e.g. the constructor):

```
uint32_t AP_Frsky_SPort_Passthrough::calc_velandyaw(void)
{
bool option_airspeed_enabled = (_frsky_parameters->_options & frsky_options_e::OPTION_AIRSPEED_AND_GROUNDSPEED) != 0;
uint32_t velandyaw = _protocol->calc_velandyaw(option_airspeed_enabled, _passthrough.send_airspeed);
// toggle air/ground speed selector
_passthrough.send_airspeed = !_passthrough.send_airspeed;
return velandyaw;
}
```

Similarly it would be for few others, like calc_param().

In short, I think this is accounted for and disentangled already.

**Comment by olliw42** [comment]
> Since you'll have exclusive use of the frsky telemetry singleton you can tune the individual streams and perhaps disable the unwanted ones

yes, this idea crossed my mind too. You'll find a respective comment in the code. 
One could actually do it either way, either a passthrough is not send if a mavlink stream/message providing that data is enabled, or vice versa, a mavlink stream/message is not send if a corresponding passthrough is send. The latter would mean that ideally GCSes also should read that mavlink passthrough. One could also make it rate dependent, e.g., if a mavlink messages is requested at a higher rate than the passthrough mavlink, one could drop the passthrough, and vice versa. So, there is lots of heuristics one can work on here.

However, I think, this can or rather should be postponed for later. IMHO one first wants to get the basics right :) I.e., get the code and structure into a mergable condition, decide if this should be a new stream SRx or linked into an existing like Extra2, decide if TUNNEL or a new message and how this new message should look like.

**Comment by olliw42** [comment]
> One open point is a good way to reuse the passthrough code

yes. :) It is spread out over several files currently

> right now we do that in the FPort/FPort2 and CRSF telemetry implementations, basically you call AP_Frsky_Telem::get_telem_data(&telem_data.packet, packet_count, 1)
The frsky telemetry is a singleton so you cannot have frsky telemetry on serial port AND an external consumer ...

I have not understood the use of external consumer and so on.

> Whenever you call that method you actually invoke the frsky WFQ packet scheduler 

I actually wanted to explicitly bypass this scheduler on purpose ... I think there is no need for it, i.e., one at most needs only very basic scheduling. I mean, it really doesn't matter much if one is sending 8 or 15 passthrough packets in that mavlink message ... it's tiny in any case compared to any other mavlink stuff ... 

I think one only wants to slow down the parameters, but this can be done with a simple last_send > 5000. And maybe some few other data could be slowed down to half rate or so.

**Comment by yaapu** [comment]
@olliw42  I agree on the priorities.
My comments were just hints on how to avoid duplicating the passthrough code with the current codebase

**Comment by yaapu** [comment]
> > One open point is a good way to reuse the passthrough code
> 
> yes. :) It is spread out over several files currently
> 
> > right now we do that in the FPort/FPort2 and CRSF telemetry implementations, basically you call AP_Frsky_Telem::get_telem_data(&telem_data.packet, packet_count, 1)
> > The frsky telemetry is a singleton so you cannot have frsky telemetry on serial port AND an external consumer ...
> 
> I have not understood the use of external consumer and so on.
> 
> > Whenever you call that method you actually invoke the frsky WFQ packet scheduler
> 
> I actually wanted to explicitly bypass this scheduler on purpose ... I think there is no need for it, i.e., one at most needs only very basic scheduling. I mean, it really doesn't matter much if one is sending 8 or 15 passthrough packets in that mavlink message ... it's tiny in any case compared to any other mavlink stuff ...
> 

if you request a multi packet with a size bigger than the number of different packet types the scheduler is alreadt bypassed

> I think one only wants to slow down the parameters, but this can be done with a simple last_send > 5000. And maybe some few other data could be slowed down to half rate or so.

**Comment by olliw42** [comment]
> avoid duplicating the passthrough code with the current codebase

my attempt was AP_Frsky_SPort_Protocol class, but I certainly may not see the full picture

in AP_Frsky_SPort_Passthrough.cpp one would do at init

 AP_Frsky_SPort_Protocol* _passthorugh;
 if (_passthrough == nullptr) {
        _passthorugh = new AP_Frsky_SPort_Protocol();
 }

Then most of the code in AP_Frsky_SPort_Passthrough.cpp could go. Not sure if this is a good or bad idea LOL.

> if you request a multi packet with a size bigger than the number of different packet types the scheduler is alreadt bypassed

as said, I am not really understanding these code parts.
Are you saying there is already all infrastructure available which one just would have to use?
I started with trying this, but failed because of cross-effects, like some internal states, and also the requirement of a valid _port, and so on. 
If there is a smarter way than what I tried, I really would need 101 explanation :)

Besides not having seen a better way, also logically it seemed not a bad idea to me to totally disentangle this. I mean, you never know, in some future someone wants to send mavlink passthrough on some telemetry serial x and at the same time wants frsky/crsf telemetry for a receiver connected to a receiver pin in the old ways ...  I think it could be a good thing to have them both, the traditional Frsky/Crsf/Fport/xxx and the mavlink passthrough work in parallel.

This are just my 2 cents and thoughts. Pl advice, decide :)

**Comment by olliw42** [comment]
AP_Frsky_SPort_Protocol could/would host functions used by
- AP_Frsky_Backend: like calc_nav_alt(), calc_gps_position(), calc_rpm(), get_vspeed_ms() and the static const uint enum's
- AP_Frsky_SPort: prep_number(), calc_gps_latlng(bool &send_latitude), and also could provide the functions for most of the sensors in AP_Frsky_SPort::send(void)
- AP_Frsky_SPort_Passthrough: all calc_xxx functions

this would clean up the code quite massively, I think

if you want, I could do that, so you could better see how it looks like ... but I could not test it beyond that it compiles, as I have no hardware setup for it

**Comment by olliw42** [comment]
so, while not knowing if I'm heading in the right direction (feedback as regards this would be useful), I moved forward and progressed in a way I would think is a good direction LOL
- used the protocol singleton to clean up the older libraries
- added a parameter SRx_FRPT
- moved to FRSKY_PASSTHROUGH_PROTOCOL instead of tunnel

I made this available now also to the mLRS and BetaCopter users (https://github.com/olliw42/mLRS/commit/91ba96dfdc89235e743b9a756787929252fc1de0, https://github.com/olliw42/BetaPilot/commit/ac52c3d5deb8ad1aac0ae04436581d1225548cbf). So, from my side it went into "production" LOL

QUESTIONS:
- why these  static const uint8_t DATA_ID_xxxx in AP_Frsky_Backend.h, and not just the usual enums?
- why many of the enums like enum PassthroughPacketType : uint8_t {...}; and not just the usual slim enum {...};?

I fail to see why that is really needed, but think these very significantly makes the code looking ugly.

My next steps probably will be to work a bit on rate managing and dropping some frsky packets depending on which other streams are enabled, so that FRSKY_PASSTHROUGH_PROTOCOL only sends what's not available as mavlink. I'm thinking also about a "high latency" bit in the mLRS protocol to allow a user to select. I expect that with such additions the code will become significantly more complex, and hence more difficult to review, and I will also become more stiff with changing direction (given I had spend so much time then) ... i.e., I suspect that with me moving further forward this here may become more and more less merge-able without steering advices ;). Anyway, that's how it is, right, earth doesn't stop rotating LOL.

**Comment by olliw42** [comment]
@yaapu 
how to make progress on this?
just some minor response from ArduPilot would be useful, and if it is just thumb down. If there is no further relevance of the PR I also can close it.
anyone else one should ping?

**Comment by yaapu** [comment]
@olliw42 I like the general idea, I would have liked better leveraging the current support for external frsky telemetry consumers rather than extrapolating all the protocol logic in a new singleton but I reckon that the latter gives much more control on how and which frames to generate.

Can't speak for the mavlink side we need the insight of other devs, @peterbarker what do you think?

**Comment by yaapu** [comment]
> QUESTIONS:
> 
> * why these  static const uint8_t DATA_ID_xxxx in AP_Frsky_Backend.h, and not just the usual enums?
> * why many of the enums like enum PassthroughPacketType : uint8_t {...}; and not just the usual slim enum {...};?

I'll try to answer letting others correct me :-)
There used to be a ton of global frsky related #define, those have been converted to const declarations which are now scoped inside the backend class.
As for enum classes a good answer and better than mine for sure is from [Bjarne Stroustrup's C++11 FAQ](http://www.stroustrup.com/C++11FAQ.html#enum)

**Comment by olliw42** [comment]
to whom ever it may concern
it might be useful to hear an opinion on whether that's of interest to ArduPilot. The actual implementation may need work, sure, but if not of interest in general, I would close, or you may feel free to close.
Thx.
